### DISCUSSION TITLE
Share Vike on Cloudflare Pages Demo?

### DISCUSSION THREAD

**Discussion by osseonews** [vikejs/vike#1403] [Discussion]
After alot of tinkering and some help from ChatGpt, I was able to get a working version of Vike on Cloudflare Pages. To help others, I would like to share this  as an example, as the current example with Cloudflare Pages is out of date and will not work as expected. How can I do this? I'm sure it's not perfect, but perhaps those are much smarter than me can work with it, to perfect it.

**Comment by brillout** [comment]
PR welcome 👍

**Reply by brillout** [reply]
Actually, I don't own the example listed at https://vike.dev/cloudflare-pages. So a new repository would be the best.

**Reply by osseonews** [reply]
I'll publish to a new repo later, and you can fork it or something, review and publish as an example. It does work, but honestly, I'm just a tinkerer so better eyes would probably improve it somewhat.

**Reply by brillout** [reply]
👍 Looking forward to it.

**Reply by osseonews** [reply]
Here is the public repo: https://github.com/osseonews/vike-app-cfp - feel free to fork and put on your website. If you can give me a shout out, that would be great (name and company is in package.json). As mentioned, I am not a full time programmer, by any means, so if you see any errors/improvements etc. please let me know. Thanks very much. BTW, we will expand on this repo over time to include a full fledged ecommerce app.

**Reply by brillout** [reply]
👍 [Done](https://github.com/vikejs/vike/commit/d485ae464c152faf91288265d382a7770f14ceca). Nice. I've add a look at the code and everything makes sense.

**Reply by brillout** [reply]
Btw. another user just told me about https://github.com/travis-r6s/vike-cf-pages.

**Reply by osseonews** [reply]
Great. Thanks for also linking to the other example. It's funny, because I considered using the middleware approach also, but didn't for exact reasons they commented on in their repo, and opted for the _worker.ts and just not accessing worker CLI in dev (actually Vite is much better in dev). Really not clear what the best approach is. For me, it was just much better to develop locally with just Express/Vite and then deploy. Really don't like Wrangler in dev. Long-term, anyway, I plan to use some routing library, like Itty or Hono, which would simplify the routing/folder structure that can occur if you use _worker.js. 

```
/**
 * This is likely best handled in a `_worker.js` file, instead of using
 * a middleware function - however, I couldn't get the worker CLI to use
 * the `_worker.ts` file in dev, and I like using the platforms conventions
 * for the functions layout/routing. If we override using a `_worker.ts` file,
 * we would need to duplicate this ourselves with a custom router, and a folder structure.
 */
```

**Reply by brillout** [reply]
FYI @travis-r6s in case you want to exchange notes with each other.

**Reply by travis-r6s** [reply]
@osseonews I'm okay with using the Miniflare based pages CLI in dev, but yeah, feels like I'm misusing the middleware a bit, although as far as I can make out it is just another worker function (no different to the other function handlers) so maybe I'm just getting hung-up over the name of it.

I did briefly look at creating a Vite plugin to handle processing the function files and building them to `/functions/[x].js` or to a `_worker.ts` so there wasn't any need to run the wrangler CLI in parallel (and you could run them with Miniflare in dev), but didn't really have the time to spend on that particular solution...

**Reply by osseonews** [reply]
@travis-r6s I just started to look at Miniflare is the idea simply that in dev mode instead of using the server/index file in Express from Vite, you replace that with the Miniflare application in dev? Do you have any examples you can share with miniflare?

**Reply by travis-r6s** [reply]
@osseonews Correct, yes - there is a plugin called [`vite-plugin-cloudflare`](https://github.com/Aslemammad/vite-plugin-cloudflare) that does this - it runs the Miniflare server in parallel with the Vite dev server, and runs the functions through it as they are requested. The benefit to this is that you are then developing your functions in the same environment as they would be in production (i.e. you can use KV, R2 etc.).

That plugin doesn't do any processing of the functions though - I'd like to update it so it passes the functions through Vite first, then runs them, so you can make use of the Vite transformations + ecosystem (Nitro does this https://nitro.unjs.io/deploy/providers/cloudflare#cloudflare-pages)

**Reply by osseonews** [reply]
@travis-r6s My problem with vite-plugin-cloudflare is that it relies on Wrangler, as does using the middleware approach. I don't like Wrangler in development for a wide variety of reasons. Anyway, I started doing some development work with Miniflare and what I discovered is that you can use Miniflare without Wrangler via a Javascript implementation. I've started building out some stuff with that, and got it to work with KV. I think I'll try this approach first, as I want to stick with the _worker.js implementation, if I can. I think it's a bit simpler and also easier to port over code from Workers to Pages if you use the workers.js mode in pages.

**Reply by travis-r6s** [reply]
@osseonews Makes sense - the company I built my demo project for wanted API routes, so I wanted to use the pages functions method.

And yes, Miniflare does indeed have an SDK you can use - an idea I had was to pass the `_worker.ts` or `/functions/*.ts` files through the Vite transformer, then pass it to Miniflare to process requests. A side project for the weekend perhaps!

**Reply by brillout** [reply]
Isn't an issue with Miniflare that it has a subpar support regarding to Cloudflare Workers specific features such as KV or Durable Objects?

**Reply by travis-r6s** [reply]
@brillout As far as I'm aware, Miniflare has support for most Cloudflare Workers platform features? E.g. https://miniflare.dev/storage/kv That being said, it seems their new v3 isn't as featureful as the previous v2.

Cloudflare is maintaining the project, and it seems to be fairly active: https://github.com/cloudflare/miniflare

**Reply by brillout** [reply]
Hm, I wonder then what the upside of using `wrangler dev` instead of Miniflare is in practice.

**Reply by travis-r6s** [reply]
@brillout I think that wrangler uses Miniflare under the hood? https://blog.cloudflare.com/wrangler3
So wrangler is in essence a CLI for Miniflare, with a few other features as well.

**Reply by osseonews** [reply]
@brillout @travis-r6s Yes, Wrangler is just Miniflare now basically. Miniflare now includes everything you need. Docs are pretty good. Benefits of Wrangler now?  I guess the main thing is that it does everything for you out of the box, while with Miniflare you need code manually? Also, I don't think Miniflare can deploy a Worker, though if you are using Cloudflare Pages this is done for you automatically when you push. Anyway, for projects like Vike, I think Miniflare on Pages will work better, though admittingly, I'm not clear how everything should be set up.

@travis-r6s For routing I prefer to just use itty-router, which is really simple. Hono is also popular, but too complex for my tastes.

**Reply by brillout** [reply]
I see, neat.

I'm glad Cloudflare is ramping up in the DX space.

If any of you would be up for a PR to create a minimal example (using `vike-react` instead of a custom `/renderer/`) at `/examples/cloudflare-workers/` using the following that'd be very much welcome 💚
 - For dev, miniflare + Travis's trick to forward all frontend requests to Vite.
 - For prod, Cloudflare Pages.

I guess we agree that's the best DX so far?

We could then link to more full-fleged examples that use itty-router or Hono.

**Reply by travis-r6s** [reply]
@brillout I'd be happy to create a simple example using `vike-react` - I'll try and work on it this w/e.

**Reply by brillout** [reply]
Neat thanks, looking forward to it 👀

Can't wait, it seems like Cloudfalre Workers is starting to get ready for prime time ✨

**Reply by osseonews** [reply]
hey all,
I updated our repo at  https://github.com/osseonews/vike-app-cfp  to use Itty Router in the _worker file, and included a sample API route to show how it could work (this API route is called in a React About page/component to make a message interactive). Note: We didn't actually finish this page yet completely, but it works as a demo. Also, included a Miniflare option in the dev environment to use KV, to show how Miniflare could work. I'll probably refactor the entire dev to use Itty Router, as you can use in Node environment, but too lazy now. Also, I think Miniflare is probably just useful for testing Cloudflare storage in dev, for everything else you can just Express or something like Itty.  Hope this helps someone. Please remember that I mainly a hobbyist, so not sure if everything is coded correctly, in terms of Vike. With Itty, I am quite familiar with it and use it on all our production sites and have been very happy with it.

**Reply by osseonews** [reply]
BTW, I would say that one of the great things about KV in Cloudflare, is that it eliminates any need for SSG. If you store your content in KV and do SSR, then that is pretty much is just as good as SSG, at least in my tests. What I've been testing is simply getting KV data in the Worker and passing to page context and then  just display on page. No need for anything else.

**Reply by brillout** [reply]
Indeed, and it seems much more powerful than Vercel's ISR.

**Comment by travis-r6s** [comment]
@brillout A simple demo of `vike-react` and Cloudflare Pages here: https://github.com/travis-r6s/vike-react-cf-pages

LMK if there is anything else I should add?

**Reply by osseonews** [reply]
@travis-r6s Thanks very much for posting this. Very helpful. What I don't understand, though, is how this works in development mode. Let's say, for example, we wanted to add something custom to pageContext and pass it to the client, the render function is not called in any server code, so how is this added in development? For production, I assume we can modify the render in the middlware file, but in development, this is not run. Also, let's say, also we need to edit the onRenderClient, how do we do that? I assume everything is handled by Vite in some way, but I'm confused how it works. I understand I can add files to onRenderClient.tsx to Pages, but then are these processed in Development? Thanks.

**Reply by brillout** [reply]
Just had a look at it, very neat. It looks promising indeed.

> What I don't understand, though, is how this works in development mode. Let's say, for example, we wanted to add something custom to pageContext and pass it to the client, the render function is not called in any server code, so how is this added in development?

Maybe there is a way to tell wrangler to forward only a certain URL pattern that matches assets URLs processed by Vite?

FYI we're working on a Vike CLI, and one thing we could do is to have an option `vike --only-vite` so that `renderPage()` isn't called (only Vite's dev server is used).

**Reply by brillout** [reply]
> Maybe there is a way to tell wrangler to forward only a certain URL pattern that matches assets URLs processed by Vite?

Actually, we can do it ourselves: we can look at the URL and if the URL matches a Vite URL pattern, then we omit `renderPage()`. I think that could work? The question here is: can we reliably distinguish Vite URLs from other URLs?

**Reply by osseonews** [reply]
Yeah, this is why I think I prefer to use miniflare and worker file, instead of functions + middleware with Vike on Cloudflare pages. It is a better dev environment, with just miniflare, i think. I've been playing around with miniflare alot and it's quite good and simple. Basically, you can mock any Cloudflare function/storage etc. by just instantiating miniflare, and then pass the result back to your server function which runs in dev (can be express). I'll try to update our public repo later this week with some examples.

**Reply by brillout** [reply]
Ok, looking forward to it 👀

**Reply by travis-r6s** [reply]
@osseonews In development, the Vite dev server can handle both Vite asset requests, and Vike SSR requests itself. So while you can run Vite + Vike together in dev mode if you need to run an API alongside it (for example in the Express demo), you can also just run `vite dev` if you have no need for an API server.

So in the demo I created, I am running the `wrangler dev` and `vite dev` commands concurrently, so the wrangler server handles any API requests, and also proxies any 'frontend' requests to the Vite dev server, which then handles all Vite asset + Vike SSR requests. 

Does that make sense?

**Reply by osseonews** [reply]
@travis-r6s Thanks. But, if we wanted to add some initial pagecontext, and we added in the _middleware when you instantiate the render for vike, will this also show up in development or only in production?

**Reply by travis-r6s** [reply]
@osseonews That is handled internally by Vike, so it doesn't really matter how Vike is instantiated - you can run it with Vite, or run `renderPage` from a server, and it should still run all the internal Vike hooks.

**Reply by osseonews** [reply]
The middeware file will run in development? (Edit: I mean the Vike specific code in middleware)

**Reply by brillout** [reply]
@travis-r6s How would you do [this](https://vike.dev/pageContext#:~:text=At%20your%20Vike%20server%20middleware%3A) then? I think that's the issue @osseonews is referring to.

The fact that `renderPage()` is hidden from the user also means that the Cloudflare `Request` object isn't available in `pageContext` and, therefore, also isn't available to the user's React/Vue components.

Most notably, the cookie/user information isn't available.

The question is: how can we, somehow, add Cloudflare's `Request` object (and/or cookie/user information) to Vike's `pageContext`?

[My previous comment](https://github.com/vikejs/vike/discussions/1403#discussioncomment-8056608) was suggesting to make `renderPage()` available to the user, so that the user can add `Request` stuff to `pageContext`. But maybe there is another way to solve this 🤔

I guess the bottom line is: how can we pass `Request` stuff to Vike?

**Reply by travis-r6s** [reply]
@osseonews No, because in development the Vite dev server handles the Vike requests. So the development request chain goes (simplified a bit):

1. Request from browser to Wrangler dev server `/about`
2. Wrangler dev server has no handler for the URL (i.e. it doesn't match `/api`), so it passes it off to the proxy
3. The proxy proxies this request to the Vite dev server `http://localhost:3000/about`
4. The Vite dev server receives this request, and Vike sees that it is a page, so takes over the rendering.
5. Vike sends back a response
6. The proxy receives this response, and sends it back to the browser

In production, the Vite dev server obviously isn't running (and we don't need it to as we already have all built assets), so the middleware file then passed the `isProduction` check, and uses Vike's `renderPage` to handle the request.

If you wanted to, you can run the `renderPage` in the middleware function even in dev, but I think I had some issues there, so I did it the way I did.

**Reply by travis-r6s** [reply]
@brillout We should simply be able to just remove the `isProduction` equivalent check in the middleware, and handle all requests with `renderPage`. I mention above though that I think I came across some issues doing it this way in dev, so it was simpler to run it the way I did. I can test this out again when I get a second though...

**Reply by brillout** [reply]
The issue is that Vike's middleware is a catch-all route (to be able to show `404` pages), so it needs to be placed after Vite's middleware. Thus my previous comment about finding a way to distinguish URLs.

> I can test this out again when I get a second though...

Neat, looking forward to it 👀. But no rush :)

**Comment by jamiehaywood** [comment]
I had a look at this, but integrating with [Hono](https://github.com/honojs/hono). I may be missing something, but how do you get Cloudflare Pages to serve pre-rendered HTML from Vike?

**Reply by osseonews** [reply]
@jamiehaywood I believe you are asking about customizing the onRenderHtml.tsx file? If so, I believe you need to use the worker.js strategy as I don't see how you can do that with standard functions strategy.  You can look at my repo linked to above which uses a worker file for Cloudflare pages. I used itty-router, but the implementation with Hono would basically be the same thing. BTW, I haven't updated my repo yet b/c I've gotten super busy on another project and haven't had time to finish all the Vike code. Hope to do it next month.

**Reply by jamiehaywood** [reply]
I was referring to serving pre-rendered HTML i.e:
```ts
// vite.config.ts
export default defineConfig({
  plugins: [
    react(),
    vike({ prerender: true }),
  ],
})
```

where pre-render outputs an `index.html` file inside a folder i.e. `/about/index.html`. It seems that the worker doesn't serve the pre-rendered HTML (I tested this by adding a comment in the html after building). Is it the case that the function just takes all requests and does SSR?

**Reply by travis-r6s** [reply]
@jamiehaywood In that case, you wouldn't need a custom worker or to instantiate Vike serverside at all, as you would just run a build, then point Cloudflare pages to your dist folder (Pages will automatically serve your static files as with any other host like Netlify etc.)...
